cli/src/commands/run.rs

```diff
@@ -24,24 +30,41 @@ use crate::ui::Ui;
 /// # Example
 ///
 /// # Run pre-commit on your local work
-/// $ jj run 'pre-commit.py .github/pre-commit.yaml' -r (main..@) -j 4
+/// $ jj run 'pre-commit run .github/pre-commit.yaml' -r (trunk()..@) -j 4
 ///
 /// This allows pre-commit integration and other funny stuff.
 #[derive(clap::Args, Clone, Debug)]
 #[command(verbatim_doc_comment)]
-pub(crate) struct RunArgs {
+pub struct RunArgs {
     /// The command to run across all selected revisions.
-    #[arg(long, short, alias = "x")]
-    command: String,
+    shell_command: String,
     /// The revisions to change.
+    /// Multiple revsets are accepted and the work will be done on a
+    /// intersection of them.
     #[arg(long, short, default_value = "@")]
     revisions: Vec<RevisionArg>,
+    /// A no-op option to match the interface of `git rebase -x`.
+    #[arg(short = 'x', hide = true)]
+    unused_command: bool,
+    /// How many processes should run in parallel, uses by default all cores.
+    #[arg(long, short)]
+    jobs: Option<usize>,
 }
 
-pub(crate) fn cmd_run(
-    _ui: &mut Ui,
-    _command: &CommandHelper,
-    _args: &RunArgs,
-) -> Result<(), CommandError> {
+pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(), CommandError> {
+    let workspace_command = command.workspace_helper(ui)?;
+    let _resolved_commits =
+        resolve_multiple_nonempty_revsets(&args.revisions, &workspace_command, ui)?;
+    let _jobs = if let Some(_jobs) = args.jobs {
+        _jobs
+    } else {
+        // Use all available cores
+
+        // SAFETY:
+        // We use a internal constant of 4 threads, if it fails
+        let available = std::thread::available_parallelism()
+            .unwrap_or(unsafe { NonZeroUsize::new_unchecked(4) });
```

Comment:
> Let's not use `unsafe`, as it makes it hard to audit unsafety, and it could introduce memory-unsafety for innocuous code changes.

I agree, but IMO it makes sense here. As this isn't user-exposed no UB can happen. I considered the `panic!()`ing variants but the `unsafe` constructor made just more sense, why pay for a `panic!()` branch when we can uphold `NonZeroUsize`s invariant.

Reply:
I might be missing something, but it looks like `_jobs` is just a `usize`. How about (if I got all the Option stuff right):

```rust
std::thread::available_parallelism().and_then(|x| x.into()).unwrap_or(4)
```

or even:

```rust
let _jobs = args.jobs.or_else(|| 
  std::thread::available_parallelism().and_then(|x| x.into())
).unwrap_or(4)
```

Reply:
I also prefer to avoid `unsafe` as long as the performance doesn't matter. The panicking bits are unreachable, and can be optimized out.

Reply:
Does not work: https://rust.godbolt.org/z/GG56c7849 see the cores() function.

Reply:
It took a bit of trial and error, but this seems to work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6c9037e8ae07a319cf3e97f0dc91aef5

Reply:
> It took a bit of trial and error on my part, but this seems to work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6c9037e8ae07a319cf3e97f0dc91aef5

Ok, I'll remove the unsafe block when I have time this week.

Reply:
In my opinion, we should forbid unsafe code/use `#![forbid(unsafe_code)]` by default. If it's not performance-critical, then we shouldn't use `unsafe` for it. For performance-critical code, we should isolate it to a single crate. The mental overhead of even acknowledging unsafe code far exceeds the cost of one branch at runtime.

Reply:
Most of our unsafe code is in iterator types where we use `std::mem::transmute()` for pretending that a reference to a value in the iterator itself is `'static`. I think there are crates that let you hide the unsafe bits for self-referential structs like that.

Reply:
The controversial unsafe usage was removed in #2545, fixing up 0 jobs is done #2553

Reply:
Oh, I didn't realize there was this whole discussion about that one :)